llvm/lib/IR/Type.cpp

```diff
@@ -214,6 +220,15 @@ int Type::getFPMantissaWidth() const {
   return -1;
 }
 
+int Type::getDFPMantissaWidth() const {
+  if (auto *VTy = dyn_cast<VectorType>(this))
+    return VTy->getElementType()->getDFPMantissaWidth();
+  assert(isDecimalFloatingPointTy() && "Not a decimal floating point type!");
+  // For both BID and DPD the width of the mantissa varies and is dependent
+  // on the combination fields.
```

Comment:
If you've got a distinction between binary floating-point and decimal floating-point types for this method, then the decimal version probably should return the width in decimal digits as opposed to integer bits. A method that returns exclusively -1 for all the types it can legally be queried on isn't a particularly useful method, IMHO.

Reply:
Agreed! This is really just a place holder. The width of the mantissa is dependent on the combination fields and the encoding, I am not quite sure at this point how to compute that.

Reply:
The values are specified in "Decimal interchange format parameters" table in section X.2.1 of [WG14 N2601](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2601.pdf) as follows:
- Decimal32: 7
- Decimal64: 16
- Decimal128: 34

My understanding is that the values are not encoding dependent.

Reply:
Thanks. I was looking for that! Why do they call it "Decimal interchange formal parameters"? Doesn't make any sense to me.

Reply:
> The values are specified in "Decimal interchange format parameters" table in section X.2.1 of [WG14 N2601](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2601.pdf) as follows:

You should probably use H.2.1 of C23, given that the TS has been incorporated into C23. (Although, very confusingly, the table is mislabeled when copied into C23).

> Thanks. I was looking for that! Why do they call it "Decimal interchange formal parameters"? Doesn't make any sense to me.

It's copy-pasted eventually from IEEE 754. Floating point types there technically aren't bound to any specific representation, and are instead based entirely on how they model parameters of b (base), p (bits/digits), Emax (largest finite exponent). The interchange formats are the formats that have a specified layout, and 7 specific layouts are given (for binary16/32/64/128 and decimal32/64/128). (They also have an additional parameter compared to the computation formats, which is the number of bits in the storage format.)

In practice, all floating point types are represented using the IEEE 754 interchange format layout on any hardware made after I was born, so "interchange format" is really synonymous with "floating point format"

Reply:
Thank you. I can see that even the table that Tom mentioned is in there!